File: strings/string_perms/string_perms.js
#!/usr/bin/node

function get_string_perms(s){
	//BC: return if len <= 1
	if (s.length <= 1){
		return [s];
	} 

	//recursive step
	else {
		var num_chars = s.length;
		var i = 0;
		var curr_perms = [];
		for (; i < num_chars; i++){
			var substr_perms = get_string_perms(s.slice(0,i) + s.slice(i+1))
			substr_perms = substr_perms.map(function(sub_perm){ return s[i] + sub_perm });
			curr_perms = curr_perms.concat(substr_perms)
		}
		return curr_perms;
	}
}

//enter desired input string as command line parameter
console.log(get_string_perms(process.argv[2] || "abc"))	

File: misc/string_perms/string_perms.py
#!/usr/bin/python
import sys

def get_string_perms(s):
    if (len(s) <= 1):
        return [s]
    else:
        curr_perms = []
        for i in range(len(s)):
            substr_perms = get_string_perms(s[:i] + s[i+1:])
            curr_perms += [s[i] + substr_perm for substr_perm in substr_perms]
        return curr_perms

if __name__ == '__main__':
	#enter desired input string as command line parameter	
	print(get_string_perms(sys.argv[1] if len(sys.argv) > 1 else "abc"))
